docs/events/2023/docker-workshop.md

```diff
@@ -0,0 +1,235 @@
+
+This makes it really great for caching. If the files that a step uses haven't changed then it can use the old version of the layer and there will be no difference.
+
+### `FROM`
+
+The `FROM` command is the first command in a docker file and it defines the base image that you are using. This is the image that you are building on top of and it is the first layer in your image.
+
+### `WORKDIR`
+
+The `WORKDIR` command defines the working directory for your application. This is the directory that your application will be run from and it is the directory that you will be in when you run commands like `RUN` and `CMD`.
+
+### `COPY`
+
+The `COPY` command copies files from your local machine into the docker image. This is how you get your application into the docker image.
+
+### `ENV`
+
+The `ENV` command sets environment variables in your docker image. This is how you can set things like the `NODE_ENV` variable in your docker image. This is useful for variables that shouldn't be included in your applications for security reasons. It's a way of passing secrets from your host machine (your computer) to the docker image.
+
+### `RUN`
+
+The `RUN` command runs a command in your docker image. This is how you install dependencies and run other commands in your docker image. Docker containers are most often build on some variant of linux so you can often use common linux commands if you need to.
+
+### `EXPOSE`
+
+The `EXPOSE` command exposes a port in your docker image. This is how you can access your application from outside the docker image. This is useful for things like web servers that need to be accessed from outside the docker image.
+
+This line in the container tells docker that the container has a program that is listening on port 3000 and that it should expose that port to the outside world.
+
+### `CMD`
+
+The `CMD` command defines the command that is run when you run the docker image. This is how you define the command that runs your application.
+
+## Advanced `Dockerfile`s
+
+Because of the way that docker images are built, you can use them to build other docker images. This is useful for things like building a frontend and then building a backend that uses the frontend. You may build your application in one docker image then copy the build into another docker image and run it.
+
+This means that the programs that you used to build the application don't need to be included in the final version of your image. This is commonly called a multi-stage build and it is a great way to reduce the size of your docker images.
+
+```dockerfile
+# Stage 1 - Base (resources required by all layers)
+FROM node:20-alpine as base
+
+WORKDIR /app
+COPY package.json yarn.lock ./
+
+# Stage 2 - Build (dependencies required by the typescript compiler)
+FROM base as build
+
+RUN yarn install --frozen-lockfile
+COPY . .
+RUN yarn build
+RUN yarn install --production --frozen-lockfile
```

Comment:
Mention that this overwrites the current node_modules with production-only node_modules. Or maybe include an explicit rm -rf ./node_modules for better illustration.